File: zig/thoth-ipc/src/demo_bounded_buffer.zig
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya and thoth-ipc contributors
//
// Cross-language bounded buffer — the classic producer/consumer problem solved
// with the byte-exact named IPC primitives.
//
// Usage (run the consumer first, then one or more producers):
//   demo_bounded_buffer consume <total>
//   demo_bounded_buffer produce <id> <count>
//
// A fixed-capacity ring lives in a shared-memory segment; access is coordinated
// by a named Mutex (so multiple producers can contend for `head`) and two
// counting Semaphores — `empty` (free slots, starts at CAP) and `full` (filled
// slots, starts at 0). Producers and the consumer can be *different languages*:
// the shm layout, the mutex and both semaphores are byte-exact across the C++,
// Rust, Swift and Zig ports.

const std = @import("std");
const ShmHandle = @import("platform/shm.zig").ShmHandle;
const Mutex = @import("sync/mutex.zig").Mutex;
const Semaphore = @import("sync/semaphore.zig").Semaphore;

const alloc = std.heap.c_allocator;

const SHM = "__BBUF__";
const MUTEX = "bbuf_m";
const EMPTY = "bbuf_e";
const FULL = "bbuf_f";
const CAP: u32 = 4;
const SLOT: usize = 48;
const SHM_SIZE: usize = 8 + CAP * SLOT;
const TMO: i128 = 10 * std.time.ns_per_s;

fn out(comptime fmt: []const u8, args: anytype) void {
    var buf: [512]u8 = undefined;
    const s = std.fmt.bufPrint(&buf, fmt ++ "\n", args) catch return;
    _ = std.c.write(1, s.ptr, s.len);
}
fn err(comptime fmt: []const u8, args: anytype) void {
    std.debug.print(fmt ++ "\n", args);
}

const Ring = struct {
    shm: ShmHandle,
    fn open() !Ring {
        var shm = try ShmHandle.acquire(SHM, SHM_SIZE, .create_or_open);
        if (shm.refCount() <= 1) { // first opener zeroes the cursors
            headPtr(&shm).* = 0;
            tailPtr(&shm).* = 0;
        }
        return .{ .shm = shm };
    }
    fn headPtr(shm: *ShmHandle) *volatile u32 {
        return @ptrCast(@alignCast(shm.ptr()));
    }
    fn tailPtr(shm: *ShmHandle) *volatile u32 {
        return @ptrCast(@alignCast(shm.ptr() + 4));
    }
    fn slot(self: *Ring, idx: u32) [*]u8 {
        return self.shm.ptr() + 8 + idx * SLOT;
    }
    fn head(self: *Ring) u32 {
        return headPtr(&self.shm).*;
    }
    fn tail(self: *Ring) u32 {
        return tailPtr(&self.shm).*;
    }
    fn setHead(self: *Ring, v: u32) void {
        headPtr(&self.shm).* = v;
    }
    fn setTail(self: *Ring, v: u32) void {
        tailPtr(&self.shm).* = v;
    }
};

fn produce(id: []const u8, count: usize) u8 {
    var ring = Ring.open() catch return 3;
    defer ring.shm.release();
    var mtx = Mutex.open(MUTEX) catch return 3;
    defer mtx.deinit();
    var empty = Semaphore.open(EMPTY, CAP) catch return 3;
    defer empty.deinit();
    var full = Semaphore.open(FULL, 0) catch return 3;
    defer full.deinit();

    var k: usize = 0;
    while (k < count) : (k += 1) {
        if (!empty.waitTimeout(TMO)) {
            err("[producer {s}] no free slot within 10s", .{id});
            return 2;
        }
        mtx.lock();
        const idx = ring.head();
        ring.setHead((idx + 1) % CAP);
        var mbuf: [SLOT]u8 = undefined;
        const body = std.fmt.bufPrint(&mbuf, "{s} #{d}", .{ id, k }) catch mbuf[0..0];
        const dst = ring.slot(idx);
        @memcpy(dst[0..body.len], body);
        dst[body.len] = 0;
        mtx.unlock();
        full.post(1);
    }
    err("[producer {s}] produced {d} items", .{ id, count });
    return 0;
}

fn consume(total: usize) u8 {
    var ring = Ring.open() catch return 3;
    defer ring.shm.release();
    var mtx = Mutex.open(MUTEX) catch return 3;
    defer mtx.deinit();
    var empty = Semaphore.open(EMPTY, CAP) catch return 3;
    defer empty.deinit();
    var full = Semaphore.open(FULL, 0) catch return 3;
    defer full.deinit();
    out("[consumer] ready — draining {d} items through a {d}-slot ring", .{ total, CAP });

    var tally = std.StringHashMap(usize).init(alloc);
    var done: usize = 0;
    var i: usize = 0;
    while (i < total) : (i += 1) {
        if (!full.waitTimeout(TMO)) {
            err("[consumer] no item within 10s after {d}/{d}", .{ i, total });
            break;
        }
        mtx.lock();
        const idx = ring.tail();
        ring.setTail((idx + 1) % CAP);
        const p = ring.slot(idx);
        const len = std.mem.indexOfScalar(u8, p[0..SLOT], 0) orelse SLOT;
        var mbuf: [SLOT]u8 = undefined;
        @memcpy(mbuf[0..len], p[0..len]);
        mtx.unlock();
        empty.post(1);
        const msg = mbuf[0..len];
        const producer = msg[0 .. std.mem.indexOfScalar(u8, msg, ' ') orelse msg.len];
        const gop = tally.getOrPut(producer) catch return 5;
        if (!gop.found_existing) {
            gop.key_ptr.* = alloc.dupe(u8, producer) catch return 5;
            gop.value_ptr.* = 0;
        }
        gop.value_ptr.* += 1;
        done += 1;
        out("[consumer] {d}/{d}  {s}", .{ i + 1, total, msg });
    }

    out("\n[consumer] summary — {d} items from {d} producer(s):", .{ done, tally.count() });
    var it = tally.iterator();
    while (it.next()) |e| out("    {s}  {d}", .{ e.key_ptr.*, e.value_ptr.* });
    return 0;
}

pub fn main(m: std.process.Init.Minimal) void {
    var storage: [8][:0]const u8 = undefined;
    var argc: usize = 0;
    var it = std.process.Args.Iterator.init(m.args);
    while (it.next()) |a| {
        if (argc >= storage.len) break;
        storage[argc] = a;
        argc += 1;
    }
    const argv = storage[0..argc];
    const parse = struct {
        fn n(s: []const u8) usize {
            return std.fmt.parseInt(usize, s, 10) catch 0;
        }
    }.n;

    if (argv.len >= 3 and std.mem.eql(u8, argv[1], "consume"))
        std.process.exit(consume(parse(argv[2])));
    if (argv.len >= 4 and std.mem.eql(u8, argv[1], "produce"))
        std.process.exit(produce(argv[2], parse(argv[3])));
    err("usage:\n  demo_bounded_buffer consume <total>\n  demo_bounded_buffer produce <id> <count>", .{});
    std.process.exit(1);
}

File: rust/libipc/src/bin/demo_bounded_buffer.rs
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR MIT
// SPDX-FileCopyrightText: 2025-2026 natyamatsya and thoth-ipc contributors
//
// Cross-language bounded buffer — the classic producer/consumer problem solved
// with the byte-exact named IPC primitives.
//
// Usage (run the consumer first, then one or more producers):
//   demo_bounded_buffer consume <total>
//   demo_bounded_buffer produce <id> <count>
//
// A fixed-capacity ring lives in a shared-memory segment; access is coordinated
// by a named `IpcMutex` (so multiple producers can contend for `head`) and two
// counting `IpcSemaphore`s — `empty` (free slots, starts at CAP) and `full`
// (filled slots, starts at 0). Producers and the consumer can be *different
// languages*: the shm layout, the mutex and both semaphores are byte-exact
// across the C++, Rust, Swift and Zig ports. (This is the classic case where a
// mutex is genuinely required — several producers mutate the same `head`.)

use std::collections::BTreeMap;

use libipc::{IpcMutex, IpcSemaphore, ShmHandle, ShmOpenMode};

const SHM: &str = "__BBUF__";
const MUTEX: &str = "bbuf_m";
const EMPTY: &str = "bbuf_e";
const FULL: &str = "bbuf_f";
const CAP: u32 = 4; // ring capacity — small, so the semaphores actually block
const SLOT: usize = 48; // fixed bytes per slot
const SHM_SIZE: usize = 8 + CAP as usize * SLOT; // head(u32) + tail(u32) + slots

/// The shared ring: `head`(u32)@0, `tail`(u32)@4, then CAP × SLOT bytes @8.
struct Ring {
    shm: ShmHandle,
}
impl Ring {
    fn open() -> Self {
        let shm = ShmHandle::acquire(SHM, SHM_SIZE, ShmOpenMode::CreateOrOpen).expect("shm");
        if shm.ref_count() <= 1 {
            // First opener zeroes the cursors (byte-exact with a fresh segment).
            unsafe {
                (shm.get() as *mut u32).write_volatile(0);
                (shm.get().add(4) as *mut u32).write_volatile(0);
            }
        }
        Self { shm }
    }
    fn head(&self) -> u32 { unsafe { (self.shm.get() as *const u32).read_volatile() } }
    fn tail(&self) -> u32 { unsafe { (self.shm.get().add(4) as *const u32).read_volatile() } }
    fn set_head(&self, v: u32) { unsafe { (self.shm.get() as *mut u32).write_volatile(v) } }
    fn set_tail(&self, v: u32) { unsafe { (self.shm.get().add(4) as *mut u32).write_volatile(v) } }
    fn slot(&self, idx: u32) -> *mut u8 { unsafe { self.shm.get().add(8 + idx as usize * SLOT) } }
}

fn main() {
    let a: Vec<String> = std::env::args().collect();
    match a.get(1).map(String::as_str) {
        Some("consume") if a.len() >= 3 => consume(a[2].parse().unwrap_or(0)),
        Some("produce") if a.len() >= 4 => produce(&a[2], a[3].parse().unwrap_or(0)),
        _ => {
            eprintln!(
                "usage:\n  demo_bounded_buffer consume <total>\n  \
                 demo_bounded_buffer produce <id> <count>"
            );
            std::process::exit(1);
        }
    }
}

fn produce(id: &str, count: usize) {
    let ring = Ring::open();
    let mutex = IpcMutex::open(MUTEX).expect("mutex");
    let empty = IpcSemaphore::open(EMPTY, CAP).expect("empty sem");
    let full = IpcSemaphore::open(FULL, 0).expect("full sem");

    for k in 0..count {
        if !empty.wait(Some(10_000)).unwrap_or(false) {
            eprintln!("[producer {id}] no free slot within 10s (consumer gone?)");
            std::process::exit(2);
        }
        mutex.lock().expect("lock");
        let idx = ring.head();
        ring.set_head((idx + 1) % CAP);
        let msg = format!("{id} #{k}");
        let b = msg.as_bytes();
        let n = b.len().min(SLOT - 1);
        unsafe {
            std::ptr::copy_nonoverlapping(b.as_ptr(), ring.slot(idx), n);
            *ring.slot(idx).add(n) = 0;
        }
        mutex.unlock().expect("unlock");
        full.post(1).expect("full.post");
    }
    eprintln!("[producer {id}] produced {count} items");
}

fn consume(total: usize) {
    let ring = Ring::open();
    let mutex = IpcMutex::open(MUTEX).expect("mutex");
    let empty = IpcSemaphore::open(EMPTY, CAP).expect("empty sem");
    let full = IpcSemaphore::open(FULL, 0).expect("full sem");
    println!("[consumer] ready — draining {total} items through a {CAP}-slot ring");

    let mut tally: BTreeMap<String, usize> = BTreeMap::new();
    for i in 0..total {
        if !full.wait(Some(10_000)).unwrap_or(false) {
            eprintln!("[consumer] no item within 10s after {i}/{total}");
            break;
        }
        mutex.lock().expect("lock");
        let idx = ring.tail();
        ring.set_tail((idx + 1) % CAP);
        let msg = unsafe {
            let p = ring.slot(idx);
            let mut len = 0;
            while len < SLOT && *p.add(len) != 0 { len += 1; }
            String::from_utf8_lossy(std::slice::from_raw_parts(p, len)).to_string()
        };
        mutex.unlock().expect("unlock");
        empty.post(1).expect("empty.post"); // free the slot
        let producer = msg.split(" #").next().unwrap_or("?").to_string();
        *tally.entry(producer).or_default() += 1;
        println!("[consumer] {:>3}/{total}  {msg}", i + 1);
    }

    println!("\n[consumer] summary — {} items from {} producer(s):", tally.values().sum::<usize>(), tally.len());
    for (p, n) in &tally {
        println!("    {p:<12} {n}");
    }
}
